---
float32 error
bool success